src/buildercore/threadbare/state.py

```diff
@@ -0,0 +1,130 @@
+import copy
+import contextlib
+
+CLEANUP_KEY = "_cleanup"
+
+
+class LockableDict(dict):
+    def __init__(self, *args, **kwargs):
+        dict.__init__(self, *args, **kwargs)
+        self.read_only = False
+
+    def update(self, new_dict):
+        if self.read_only:
+            raise ValueError(
+                "dictionary is locked attempting to `update` with %r" % new_dict
+            )
+        dict.update(self, new_dict)
+
+    def __setitem__(self, key, val):
+        # I suspect multiprocessing isn't copying the custom 'read_only' attribute back
+        # from the child process results. be aware of this weirdness
+        # print("self:", self.__dict__, "internal data:", self)
+        if hasattr(self, "read_only") and self.read_only:
+            raise ValueError(
+                "dictionary is locked attempting to `__setitem__` %r with %r"
+                % (key, val)
+            )
+        dict.__setitem__(self, key, val)
+
+
+def read_only(d):
+    if hasattr(d, "read_only"):
+        d.read_only = True
+
+
+def read_write(d):
+    if hasattr(d, "read_only"):
+        d.read_only = False
+
+
+def initial_state():
+    """returns a new, empty, locked, LockableDict instance that is used as the initial `state.ENV` value.
+
+    if you are thinking "it would be really convenient if 'some_setting' was 'some_value' by default",
+    see `set_defaults`."""
+    new_env = LockableDict()
+    read_only(new_env)
+    return new_env
+
+
+ENV = initial_state()
+
+DEPTH = 0  # used to determine how deeply nested we are
+
+
+def set_defaults(defaults_dict=None):
+    """re-initialises the `state.ENV` dictionary with the given defaults.
+    with no arguments, the global state will be reverted to it's initial state (an empty LockableDict).
+
+    use `state.set_defaults` BEFORE using ANY other `state.*` functions are called."""
+    global ENV, DEPTH
+    if DEPTH != 0:
+        msg = "refusing to set initial `threadbare.state.ENV` state within a `threadbare.state.settings` context manager."
+        raise EnvironmentError(msg)
+
+    new_env = LockableDict()
+    new_env.update(defaults_dict or {})
+    read_only(new_env)
+    ENV = new_env
+
+
+def cleanup(old_state):
+    if CLEANUP_KEY in old_state:
+        for cleanup_fn in old_state[CLEANUP_KEY]:
+            cleanup_fn()
+        del old_state[CLEANUP_KEY]
+
+
+def _add_cleanup(state, fn):
+    cleanup_fn_list = state.get(CLEANUP_KEY, [])
+    cleanup_fn_list.append(fn)
+    state[CLEANUP_KEY] = cleanup_fn_list
+
+
+def add_cleanup(fn):
+    "add a function to a list of functions that are called after leaving the current scope of the context manager"
+    return _add_cleanup(ENV, fn)
```

Comment:
needs a better differentiation from `add_cleanup()` in terms of naming

Reply:
why?

Reply:
not "why do we need to give functions clear names", but "why does the wrapper need to clearly differentiate itself from it's wrapped private function?".

it seems clear as crystal to me

Reply:
Not really sure what the separation is doing as this seems to be the only call to `_add_cleanup` and given the `_` I'd consider it private to this module :thinking: Nitpicking, do not want to change it either way

Reply:
it follows a general pattern I've taken on with Clojure where side effecting code is pushed to the fringes, typically because testing in Clojure is slightly harder than in Python so the less scaffolding, the better.

In this case however my instincts are wrong as the 'pure' function is modifying a global state dict and not returning anything. This module underwent a lot of churn late in the process as I removed a big feature and added `set_defaults`. I'll take a second look at it today as I fix up the py2 typing issue breaking CI